# Arrays and Matrices 

## Learning Objectives 
- Understand the concept of arrays and matrices in Julia 
- Create and initialise one-dimensional and multi-dimensional arrays 
- Utilise various array creation functions in Julia (e.g. `zeros`, `ones`, random arrays)
- Perform basic manipulations such as indexing and slicing on arrays
- Implement common operations on arrays and matrices

Arrays are a fundamental data structure in Julia that is used to store collections of elements. If you're coming from Python, you can think of one-dimensional Julia arrays as similar to Python lists or NumPy 1D arrays, and multi-dimensional Julia arrays are like NumPy arrays or matrices. One key difference: Julia arrays are **1-indexed** (the first element is index 1, not 0). 

In Julia, all elements of a typical array have the same type (e.g. an `Array{Int}` or 'Array{Float64}`), which allows for performance optimisations. It is possible to have arrays of mixed types (they would have element type `Any`), but most often, we use uniform-type arrays. 

## One-Dimensional Arrays (Vectors) 
A one-dimensional array (also called a **vector**) is essentially an ordered list of elements. The simplest way to create one is by using square brackets `[...]` with commas separating elements: 

In [4]:
# Declare and initialize a 1D array of integers
a = [1, 2, 3]
println(a)  
println(typeof(a))   # e.g., Output: Vector{Int64} (which is an alias for Array{Int64,1})


[1, 2, 3]
Vector{Int64}


The type `Vector{Int64}` indicates a one-dimensional array (vector) of 64-bit integers. Julia infers the type from the elements you provided. If you mix types, Julia will promote to a common type if possible or else use a generic type. For example, `[1, 2.5]` would result in a vector of `Float64` (because one can be promoted to 1.0 to match 2.5), but `[1, "two"]` would result in a vector of type `Any` (mixed types, generally to be avoided for performance). 

You can also create an empty array and fill it later:

In [5]:
empty_vec = Int[]  # an empty Vector of Int
push!(empty_vec, 10)
push!(empty_vec, 20)
println(empty_vec)  


[10, 20]


Here, we used `push!` to append elements to the dynamic array. 

## Array Constructors 
Julia provides an `Array` constructor for more control, especially if you want to create uninitialised arrays or with specific element types: 



In [8]:
# Create an uninitialized 1D array (vector) of length 3 for Int elements
b = Array{Int}(undef, 3)
println(b)


[71, 74, 5369377792]


The `undef` keyword tells Julia we want to allocate the array without initialising its contents (for performance). The contents of `b` will be whatever was already in those memory locations (essentially garbage or random-looking numbers). You should assign values to all positions before using them, or else those initial values are meaningless. The benefit of `undef` is if you plan to populate the array immediately, you don't spend time setting initial values. 

For instance, if we print `b` right after creation, we might see something like `[71, 74, 5369377792]`; these are just whatever bits were in memory. After we populate it: 

In [9]:
b[1] = 10; b[2] = 20; b[3] = 30;
println(b)  # Now it will reliably print [10, 20, 30]

[10, 20, 30]


Usually, if you have initial values, it's simpler to use the bracket notation as we did with `a`. The `Array` constructor is more often used to create large arrays for efficiency. 

Julia also offers convenient functions to create arrays filled with specific values: 
- `zeros(n) - a length-`n` vector of all zeros (as Float64 by default, or you can call `zeros(Int, n)` for Int zeros. 
- `ones(n) - a length-`n` vector of all ones.
- `rand(n) - a length-`n` vector of random numbers uniformly distributed in [0,1). 
- `randn(n) - a length-`n` vector of random numbers from a normal (Gaussian) distribution. 

For example: 

In [10]:
println( zeros(5) )   # e.g., [0.0, 0.0, 0.0, 0.0, 0.0]
println( ones(Int, 4) )  # e.g., [1, 1, 1, 1] as Int64
println( rand(3) )    # e.g., [0.435, 0.075, 0.912] (3 random floats)
println( randn(3) )   # e.g., [1.2, -0.5, 0.3] (3 random normal floats)

[0.0, 0.0, 0.0, 0.0, 0.0]
[1, 1, 1, 1]
[0.09821129017798369, 0.6011556206492765, 0.37222564056165997]
[-2.081913081273531, -0.5654357053623454, 0.19193288869427932]


## Multi-Dimensional Arrays (Matrices)
Multi-dimensional arrays are essentially matrices (2D) or higher-dimensional tensors. You can create a 2D array (matrix in Julia using semicolons `;` to separate rows within the square brackets: 

In [11]:
# 2x3 matrix (2 rows, 3 columns)
M = [1 2 3; 
     4 5 6]
println(M)


[1 2 3; 4 5 6]


In the above, the first row is `1 2 3`, and the second row is `4 5 6`. Julia interprets that as a 2-row, 3-column matrix of Ints. 

You can use the `Array` constructor similarly for multi-d arrays. For example:

In [12]:
N = Array{Float64}(undef, 2, 3)  # 2x3 uninitialized matrix of Float64
println(N)

[5.0e-324 0.0 5.0e-324; 0.0 5.0e-324 0.0]


`N` will be a 2x3 matrix with arbitrary initial values. You would typically fill it in or use functions like `zeros(2,3)` for a zero matrix, etc. 

All the earlier initialisation functions have multi-dimensional forms: 

In [13]:
println( zeros(3,3) )   # 3x3 matrix of 0.0
println( ones(2,4) )    # 2x4 matrix of 1.0
println( rand(2,2) )    # 2x2 matrix of uniform random numbers
println( randn(3,2) )   # 3x2 matrix of normal random numbers

[0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0]
[1.0 1.0 1.0 1.0; 1.0 1.0 1.0 1.0]
[0.8825747481551981 0.6376257660252757; 0.8669537217566343 0.27277960116307]
[-0.2588904726728553 0.3233850669860786; -1.1911815060808717 0.3389154204642444; -0.18151761217880635 0.8948282370255342]


These will create matrices of the given shape filled with the specified content. 

## Exercise: Creating Matrices 
Create a 3x3 identity matrix manually using the semicolon notation; the identity matrix has 1s on its diagonal and 0s elsewhere). Verify its structure by printing it. 

## Indexing and Slicing 
Manipulating arrays involves accessing and modifying their elements. In Julia: 
- Array **indices start at 1**. This is crucial to remember (if you try to access index 0, you'll get a BoundsError). 
- Use square brackets to index. For a vector `v`, `v[1]` is the first element, `v[2]` the second, etc. For a matrix `M`, use `M[i,j]` to access the element at row i, column j. 

Example (1D indexing): 

In [15]:
v = [10, 20, 30, 40, 50]
println( v[3] )   # prints 30, the 3rd element
v[3] = 35         # modify the 3rd element
println( v )      # now v is [10, 20, 35, 40, 50]


30
[10, 20, 35, 40, 50]


Example (2D indexing):

In [16]:
M = [10 20 30 40 50;
     60 70 80 90 100]
println( M[2, 3] )  # row 2, col 3 -> this is 80 in the matrix above
M[1, 5] = 55        # set the element in row 1, col 5 to 55
println(M)

80
[10 20 30 40 55; 60 70 80 90 100]


Julia also supports `slicing` to get subarrays. You can use the `:` operator to indicate a range of indices. For instance: 
- `v[2:4]` gives a subarray consisting of the 2nd through 4th elements of vector `v`. Note: The result is a copy of those elements into a new array. 
- In a matrix, you can slice along each dimension: `M[1:2, 2:4]` would give the submatrix that spans rows 1 to 2 and columns 2 to 4. 
- A single colon `:` can mean "all indices in this dimension". For example, `M[:, 1]` gives the first column of `M` (as a vector), and `M[2, :]` gives the second row of `M` (as a 1-row matrix or a vector - Julia will preserve 1-dimensional shape for a row unless you specifically convert it). 

For example: 

In [18]:
v = [10, 20, 30, 40, 50]
subv = v[2:4]            # this will be [20, 30, 40]
println(subv)

M = [1 2 3; 
     4 5 6;
     7 8 9]
println( M[1:2, 2:3] )   # submatrix of first 2 rows and cols 2-3 -> [[2 3]; [5 6]]
println( M[:, end] )     # all rows, last column -> [3, 6, 9]^T (as a vector of length 3)


[20, 30, 40]
[2 3; 5 6]
[3, 6, 9]


Here, we introduce the special index `end`, which is a convenient way to refer to the last index of that dimension. In `M[:, end]`, `end` represents the last column index (which is 3 for our 3x3 example). 

You can also modify subarrays by slicing on the left-hand side of an assignment: 


In [19]:
v[2:4] = [100, 200, 300]   # replace a segment of v with new values
println(v)  # if v was [10,20,30,40,50], now it becomes [10, 100, 200, 300, 50]


[10, 100, 200, 300, 50]


## Common Operations on Arrays 
Some common things you might do with arrays: 
- **Append elements**: Use `push!(array, value)` to append a single value to the end of a 1D array (vector). Use `append!(array, collection)` to attend all elements of another collection. 
- **Remove elements**: Use `pop1(array`) to remove the last element (and return it). There's also `deleteat!(array, index)` to remove the element at a specific index. 
- **Combining arrays**: You can concatenate arrays using `vcat`(vertical concatenation) and `hcat` (horizontal concatenation) functions, or simply using brackets with semicolons and spaces appropriately. For example, `c = [a; b]` will stack vector `a` on top of `b` (if they have the same length), and `d = [ a b]` will put `a` and `b` side by side (if they are the same length and are column vectors). 
- **Element-wise operations**: By default, arithmetic on arrays with the same dimensions is element-wise in Julia (unlike MATLAB, which does matrix multiplication by default with `*`). For example, `u = [1,2,3]; v = [10,10,10]; println(u + v)` yields `[11,12,13]`. However, note that `*` for matrices/vector does matrix multiplication (not elementwise). For elementwise multiplication or division on arrays, use the dot form: `u .* v`, `u ./ v`, etc. 
- **Built-in functions**: Julia has many handy functions like `sum(array)` to sum elements, `maximum(array) / minimum(array)`, `sort(array)`, etc. 

## Exercise: Operations on Arrays
Given an array `data = [5, 3, 8, 1, 2]`, do the following: 
- Sort the array (ascending) and print the sorted result. 
- Add a new number (e.g. 10) to the end of the array. 
- Compute the sum of all elements in the array. 
- Create a new array `squares` that is the element-wise squares of `data` 

In [24]:
using JSON

function show_quiz_from_json(path)
    quiz_data = JSON.parsefile(path)

    html = """
    <style>
    .quiz-question {
        background-color: #6c63ff;
        color: white;
        padding: 12px;
        border-radius: 10px;
        font-weight: bold;
        font-size: 1.2em;
        margin-bottom: 10px;
    }

    .quiz-form {
        margin-bottom: 20px;
    }

    .quiz-answer {
        display: block;
        background-color: #f2f2f2;
        border: none;
        border-radius: 10px;
        padding: 10px;
        margin: 5px 0;
        font-size: 1em;
        cursor: pointer;
        text-align: left;
        transition: background-color 0.3s;
        width: 100%;
    }

    .quiz-answer:hover {
        background-color: #e0e0e0;
    }

    .correct {
        background-color: #4CAF50 !important;
        color: white !important;
        border: none;
    }

    .incorrect {
        background-color: #D32F2F !important;
        color: white !important;
        border: none;
    }

    .feedback {
        margin-top: 10px;
        font-weight: bold;
        font-size: 1em;
    }
    </style>

    <script>
    function handleAnswer(qid, aid, feedback, isCorrect) {
        // Reset all buttons for the question
        let buttons = document.querySelectorAll(".answer-" + qid);
        buttons.forEach(btn => {
            btn.classList.remove('correct', 'incorrect');
        });

        // Apply correct/incorrect to selected
        let selected = document.getElementById(aid);
        selected.classList.add(isCorrect ? 'correct' : 'incorrect');

        // Show feedback below the question
        let feedbackBox = document.getElementById('feedback_' + qid);
        feedbackBox.innerHTML = feedback;
        feedbackBox.style.color = isCorrect ? 'green' : 'red';
    }
    </script>
    """

    for (i, question) in enumerate(quiz_data)
        qid = "$i"
        html *= """<div class="quiz-question">$(question["question"])</div><form class="quiz-form">"""

        for (j, answer) in enumerate(question["answers"])
            aid = "q$(i)_a$(j)"
            feedback = answer["feedback"]
            correct = startswith(lowercase(feedback), "correct")
            html *= """
            <button type="button" class="quiz-answer answer-$qid" id="$aid"
                onclick="handleAnswer('$qid', '$aid', '$feedback', $(correct))">
                $(answer["answer"])
            </button>
            """
        end

        html *= """<div class="feedback" id="feedback_$qid"></div></form><hr>"""
    end

    display("text/html", html)
end


# Use the function
show_quiz_from_json("questions/summary_arrays_and_matricies.json")

What is the result of the expression v[2:4] in Julia, given v = [10, 20, 30, 40, 50]? 
 [10, 30, 50]
 
 
 [20, 30, 40]
 
 
 [2, 3, 4]
 
 
 [30, 40, 50]
 
 Which of the following correctly creates a 2x3 matrix in Julia? 
 M = [1, 2, 3; 4, 5, 6]
 
 
 M = [[1, 2, 3]; [4, 5, 6]]
 
 
 M = [1 2 3; 4 5 6]
 
 
 M = [1; 2; 3; 4; 5; 6]
 
 How do you perform element-wise multiplication between two arrays u and v in Julia? 
 u * v
 
 
 dot(u, v)
 
 
 u .^ v
 
 
 u .* v